File: _experimental/_clojure/words/src/numerical.clj
(ns numerical)

(defn fibonacci [n]
  (cond
    (< n 0) -1
    (< n 2) n
    :else
    (loop [n n
           acc2 0
           acc1 1]
      (if (<= n 2)
        (+ acc1 acc2)
        (recur (dec n) acc1 (+ acc1 acc2))))))

(defn factorial [n]
  (if (< n 0)
    -1
    (loop [n n
           acc 1]
      (if (<= n 1)
        acc
        (recur (dec n) (* n acc))))))

(defn gcd [a b]
  (loop [a a
         b b]
    (if (= b 0)
      a
      (recur b (mod a b)))))

(defn lcm [a b]
  (if (or (= a 0) (= b 0))
    0
    (* (/ a (gcd a b)) b)))

File: _experimental/_ocaml/words/lib/numerical.ml
let fibonacci n =
  let rec fib_iter n acc2 acc1 =
    if n < 0 then -1
    else if n <= 0 then n
    else if n == 2 then acc1 + acc2
    else fib_iter (n - 1) acc1 (acc1 + acc2)
  in
  fib_iter n 0 1

let factorial n =
  let rec fact_iter n acc =
    if n < 0 then -1
    else if n <= 1 then acc
    else fact_iter (n - 1) (n * acc)
  in
  fact_iter n 1

let rec gcd a b =
  if b = 0 then a
  else gcd b (a mod b)

let lcm a b =
  (a / (gcd a b)) * b